Add 'zoom-box' to the viewer

lukasz-migas opened the pull request:
# References and relevant issues

This PR adds an interactive zoom box to the viewer canvas. 
This closes the original issue #7630 

# Description

It's only active in 2-D view and all you need to do is hold Shift, press-down in the canvas, drag the mouse and release. Once released, a `zoom` event is emitted by the `ViewerModel` which is processed in the `canvas` class.

What's currently handled - things that need addressing by me:
- [x] add support for changing axes order
- [x] add controls for how the box is displayed - currently hard-coded in the vispy code
- [x] improve how the zoom level is calculated - unlike in eg matplotlib, we need to calculate the center of zoom and zoom scale. I found this to be slightly harder to estimate so I've got code that kinda works but could definitely do with improvement.
- [x] add tests  

https://github.com/user-attachments/assets/789407d6-aa4d-48d1-9938-df95e59ed118

lukasz-migas committed:
Add

lukasz-migas committed:
Improvements

- updated zoom method
- updated line plotting which now takes width/color from model
-

lukasz-migas committed:
Update zoom.py

brisvag commented on src/napari/_vispy/overlays/zoom.py:
```diff
@@ -0,0 +1,69 @@
+                        bounds = vb
+                    elif vb is not None:
+                        bounds = [min(bounds[0], vb[0]), max(bounds[1], vb[1])]
+        return bounds
```
Nice, this and the corresponding `reset()` part can just go to the superclass I think!

lukasz-migas replied:
You're right; I had that in because I initially made it work in my own library. Will do.

brisvag commented on src/napari/components/_viewer_mouse_bindings.py:
```diff
@@ -38,3 +38,32 @@ def double_click_to_zoom(viewer, event):
+        yield
+
+    # on mouse move
+    while event.type == 'mouse_move' and 'Shift' in event.modifiers:
```
To tell other listeners we took care of this event:

```suggestion
        yield

	event.handled = True
	
    # on mouse move
    while event.type == 'mouse_move' and 'Shift' in event.modifiers:
```

brisvag commented on src/napari/components/viewer_model.py:
```diff
@@ -306,6 +310,10 @@ def scale_bar(self):
+    @property
+    def zoom(self):
+        return self._overlays['zoom']
+
```
I don't think this should be ever accessed publicly. I would make private and call `_zoom_box` for clarity.

brisvag commented on src/napari/components/viewer_model.py:
```diff
@@ -260,7 +263,7 @@ def __init__(
         # Add extra reset_view event. Ideally this should be removed in the
         # future.
-        self.events.add(reset_view=Event)
+        self.events.add(reset_view=Event, zoom=Event)
```
I think I would prefer this event to live in the overlay, by adding something like this in `ZoomOverlay`:

```py
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._events.add(zoom=Event)
```

lukasz-migas replied:
Sure! 👍

brisvag commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,30 @@
+    x_max: float,
+    y_min: float,
+    y_max: float,
+    viewer: ViewerModel,
```
Here if we want to go 3D, we should be able to use a similar logic to these lines in `QtWindow._screenshot`:

https://github.com/napari/napari/blob/d5fa476d51910301f2c18c5857f7f002cc1ea70e/src/napari/_qt/qt_main_window.py#L1701-L1719

Bassically we can get the current 2D bounding box in canvas space and zoom according to that.

brisvag reviewed (commented):
Looking good! Left a few comments :)

lukasz-migas committed:
Apply suggestions from code review

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

lukasz-migas committed:
Implemented brisvag's comments

- merged InteractionBox code
- migrated zoom from viewer to zoom_box
- changed visibility of `zoom` to `_zoom_box`

lukasz-migas committed:
Improvements

- changed to using Meta key
- started work on adding 3d support

lukasz-migas committed:
Update _viewer_mouse_bindings.py

lukasz-migas committed:
Update zoom.py

lukasz-migas committed:
Added a few tests

lukasz-migas committed:
Fix zoom in multiscale images

lukasz-migas committed:
Update qt_layer_controls_base.py

lukasz-migas committed:
Update zoom.py

lukasz-migas committed:
No longer using overlay color/thickness

lukasz-migas committed:
Few fixes

lukasz-migas committed:
This commit is broken

lukasz-migas committed:
Still broken..

lukasz-migas committed:
Fixes

lukasz-migas committed:
More tests

lukasz-migas committed:
Cleanup

- remove data_positions from the overlay
- removed tmp code
- updated tests

lukasz-migas committed:
Fixes

lukasz-migas committed:
fix tests

TimMonko commented on src/napari/_qt/layer_controls/qt_layer_controls_base.py:
```diff
@@ -107,7 +108,10 @@ def __init__(self, layer: Layer) -> None:
-                '\n(or hold Space)\n(hold Shift to pan in 3D)'
+                '\n(or hold Space)'
+                '\n(hold {control} to zoom-in using region of interest)'
+                '\n(hold Shift to pan in 3D)',
```
```suggestion
                '\n(or hold Space)'
                '\n(hold Shift to pan in 3D)',
                '\n(hold {control} to zoom via ROI selection)'
```
I'm trying to think of something to make it more clear that this isn't using any already present ROI on the interface, such as a shapes layer, but is rather done using an action. 
I also would prefer this added to the end of the tooltip, but that's just a nit

TimMonko commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,78 @@
+    # calculate average zoom based on the size of the rectangle
+    zoom = np.min(total_size / spread)
+    scale_factor = viewer._get_scale_factor(0.05)
+    if viewer.dims.ndisplay == 3:
```
```suggestion
    if viewer.dims.ndisplay == 2:
```
ok, so I'm guessing this must be switched, but fixing this behavior isn't fixing the not-quite-as-expected zooming in 3D mode. The calculation isn't going as expected somewhere when the angles are far from the default, but it _is_ using the calculation I implemented so it might be my fault. I'm getting a bit tired, so this is also a not-to-self to look at it in more detail when I have time.

TimMonko commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,78 @@
+    # adjust zoom by the native zoom factor
+    zoom = zoom * native_zoom
+    # ensure zoom is a valid number
+    if np.isinf(zoom) or np.isnan(zoom) or zoom < 1:
```
why is this clamped at 1? AFAIK it just needs to be > 0

lukasz-migas replied:
If the value is less than 1, then it will start zooming out, which I don't think is desired.

brisvag replied:
In what cases can it be < 1? Does that only happen if you somehow make a box that's bigger than the canvas?

TimMonko replied:
There are many real instances where zoom can be less than one (scale is one of them, currently so is grid mode, but that's changing). So we just want to clamp <=0
```
import napari

viewer = napari.Viewer()
viewer.open_sample('napari', 'astronaut')
viewer.layers[0].scale = (10,10)
viewer.camera.zoom = 1'
```
![image](https://github.com/user-attachments/assets/9044d619-0be2-4681-9fa7-34374b2b9df1)

TimMonko replied:
```suggestion
    if np.isinf(zoom) or np.isnan(zoom) or zoom <= 0:
```

TimMonko replied:
Oh, and it's also because I want to be able to zoom out if I draw the box like that lol, maybe that's just me though...

TimMonko commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,78 @@
+    center, spread = _get_dim_info(mins, maxs)
+
+    # calculate average zoom based on the size of the rectangle
+    zoom = np.min(total_size / spread)
```
`C:\Users\timmo\Documents\GitHub\napari\src\napari\_vispy\utils\zoom.py:57: RuntimeWarning: divide by zero encountered in divide!`

lukasz-migas replied:
Ok, this should be resolved now.

TimMonko commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,78 @@
+    else:
+        native_zoom = viewer._get_3d_camera_zoom(
+            extent, total_size, scale_factor
+        )
```
```suggestion
	selected_extent = np.vstack((mins, maxs))
        native_zoom = viewer._get_3d_camera_zoom(
            selected_extent, spread, scale_factor
        )
```
This fixes 3D camera zoom calculation to use the selection box for the native_zoom, but then I'm confused because restricting the 2D calculation like this doesn't work right.

lukasz-migas committed:
Update src/napari/_qt/layer_controls/qt_layer_controls_base.py

Co-authored-by: Tim Monko <timmonko@gmail.com>

lukasz-migas committed:
Update zoom.py

lukasz-migas committed:
Update qt_layer_controls_base.py

lukasz-migas committed:
Fix annotations

lukasz-migas committed:
Merge branch 'main' into add-zoom-box

TimMonko committed:
Merge branch 'main' into add-zoom-box

TimMonko committed:
fix 3d calc change that snuck through

brisvag commented on src/napari/_vispy/canvas.py:
```diff
@@ -175,11 +176,15 @@ def __init__(
         self.viewer.layers.events.removed.connect(self._remove_layer)
+        # Connecting events from Overlays
+        self.viewer._zoom_box.events.zoom.connect(self._on_boxzoom)
         self.viewer._overlays.events.added.connect(
```
Sorry to be nitpicky, but I find this kind of comments just noisy when reading :sweat_smile: the code is perfectly clear by itself. Blank lines in between would do a better job.

brisvag commented on src/napari/_vispy/canvas.py:
```diff
@@ -341,6 +346,16 @@ def _on_interactive(self) -> None:
+            event.value, self.viewer
+        )
+        self.viewer.camera.center = (z_center, x_center, y_center)
+        self.viewer.camera.zoom = zoom
```
codecov marks this stuff as untested, we should probably cover it!

lukasz-migas replied:
Added a test that covers this now.

brisvag commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,81 @@
+
+def _get_dim_info(
+    mins: np.ndarray, maxs: np.ndarray
+) -> tuple[np.ndarray, np.ndarray]:
```
I would rename this to something more explicit (like `_get_zoombox_center_and_size`)

brisvag commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,81 @@
+def _get_data_extents(
+    data_positions: tuple[tuple[float, ...], tuple[float, ...]],
+    displayed: tuple[int, ...],
+) -> tuple[np.ndarray, np.ndarray]:
```
Similarly, I would rename this to something like `_zoombox_corners_world_to_displayed` (assuming I understand correctly this function)

brisvag commented on src/napari/_vispy/visuals/interaction_box.py:
```diff
@@ -105,3 +111,15 @@ def set_data(
+        bounds = None
+        with suppress(ValueError):
+            for v in view._subvisuals:
+                if v.visible:
```
We need a comment about *what* is being suppressed, when it can happen, and why it't not handled but just suppressed. Also, can this suppress be restricted to a smaller portion of code?

lukasz-migas replied:
Will add a comment regarding < 1 zoom. It actually only happens in 3-D cases when the zoom box is quite large (>50% of the canvas).

lukasz-migas replied:
> We need a comment about _what_ is being suppressed, when it can happen, and why it't not handled but just suppressed. Also, can this suppress be restricted to a smaller portion of code?

Good point - I remember having to add this suppression a while back in my own codebase, but helpfully I didn't leave a comment for myself when it happened. I am gonna try to reproduce it and will add a comment here.

lukasz-migas replied:
I think the ValueError only appeared when the viewer had no data and the bounds were not specified properly. My initial implementation wasn't using the new API to get the canvas extents which probably meant I was using incorrect values. I think it's safe to remove this code and I've just done so.

lukasz-migas committed:
Fix few naming issues

lukasz-migas committed:
Update test_viewer_mouse_bindings.py

lukasz-migas committed:
Update interaction_box.py

lukasz-migas committed:
Update names

jacopoabramo commented on src/napari/_vispy/overlays/zoom.py:
```diff
@@ -0,0 +1,44 @@
+
+        self.overlay.events.canvas_positions.connect(self._on_positions_change)
+
+        self._on_visible_change()
```
@brisvag I'm wondering if it would be worth considering the option of a `__post_init__` hook to avoid having to call `__init__` where it's not necessary; looks to me that this overlay might be inherited directly from `VispySelectionOverlay`, or am I mistaken?

lukasz-migas replied:
Do you mean the `_VispyBoundingBoxOverlay` in `interaction_box.py`? That uses the `VispySceneOverlay` whereas this PR uses `VispyCanvasOverlay` so perhaps wouldn't work.

jacopoabramo replied:
Ahhhhh nevermind then, it's too late for me to elaborate reasonable comments @.@ I need to eat

jacopoabramo commented on src/napari/components/overlays/zoom.py:
```diff
@@ -0,0 +1,39 @@
+        cls, v: tuple[tuple[float, ...], tuple[float, ...]]
+    ) -> tuple[tuple[float, float, float], tuple[float, float, float]]:
+        tup_1, tup_2 = v
+        return tuple(tup_1), tuple(tup_2)
```
```suggestion
        return tuple(*tup_1), tuple(*tup_2)
```

lukasz-migas replied:
This will actually throw a TypeError since tuples take a single argument. 🍽️ 😄

jacopoabramo commented on src/napari/_vispy/overlays/zoom.py:
```diff
@@ -0,0 +1,44 @@
+class VispyZoomOverlay(ViewerOverlayMixin, VispyCanvasOverlay):
+    """Zoom box overlay.."""
+
+    def __init__(self, viewer, overlay: ZoomOverlay, parent=None):
```
```suggestion
    def __init__(self, viewer, overlay, parent=None):
```

lukasz-migas committed:
fixes to mypy and tests

lukasz-migas committed:
Merge branch 'main' into add-zoom-box

lukasz-migas committed:
Update src/napari/_vispy/utils/zoom.py

Co-authored-by: Tim Monko <timmonko@gmail.com>

lukasz-migas committed:
Added type annotations

lukasz-migas committed:
Update zoom.py

TimMonko commented on src/napari/components/_viewer_mouse_bindings.py:
```diff
@@ -38,3 +38,39 @@ def double_click_to_zoom(viewer, event):
+
+
+def drag_to_zoom(viewer, event):
+    """Enable zoom."""
```
Another nice spot to improve the docstring to describe what the function should be doing. Doesn't have to be much, and I know the examples in this file are non-existent lol

lukasz-migas replied:
Lol, done :D

TimMonko commented on src/napari/components/_viewer_mouse_bindings.py:
```diff
@@ -38,3 +38,39 @@ def double_click_to_zoom(viewer, event):
+    # only trigger zoom if the box is larger than a pixel
+    distance = np.abs(np.array(press_pos) - np.array(move_pos))
+    if distance.min() > 10:
+        viewer._zoom_box.events.zoom(value=(press_position, move_position))
```
I think this distance guard only makes sense if scale is 1.0 and zoom is roughly normally.
Here is a quick gif on what I would see people think is 'reasonable' still and it not actually cause a zoom. What's a different implementation we could try here? It's also acceptable to merge as is and come back if people have feedback imo. 
![python_h2sshAPafF](https://github.com/user-attachments/assets/20e55e9a-f5ca-46fa-aa2e-cca45efb3aa6)

lukasz-migas replied:
Technically, this condition could be disabled with no safeguards in place. I thought from a practical point of view, it's rare that a user would want such a small zoom, but that could just be me. I've made a slight change to the code to use a constant with a better explanation of what it is, and also adjusted the value down to 5px.

TimMonko commented on src/napari/_vispy/utils/zoom.py:
```diff
@@ -0,0 +1,85 @@
+    else:
+        native_zoom = viewer._get_3d_camera_zoom(
+            np.vstack((mins, maxs)), spread, scale_factor
+        )
```
ok, so what's been holding me up is that 2D uses `total_size` for the extent, whereas 3D uses the zoombox extents. 
However, this _works_ and I haven't figured out in my brain why. I'm guessing its a difference in how I actually implemented `_get_2d_camera_zoom` but it's still not clicking. 
So really, this was the only thing holding up my approval, and that's just because I wanted to get it right. I still can't figure it out, so now I'll just approve.

lukasz-migas replied:
I do think it's down to the implementation of the 3-D zoom calculation. The 3-D function seems to use the total_size in the calculation, but I haven't dug too deep into what happens under the hood...

brisvag replied:
NOTE: this *WILL* break when merging #7870 - which is likely to happen just before/just after this -- because the signature of these functions changed. It's also likely to no cause conflicts but simply fail, so we need to keep an eye out, and hopefully some tests will fail.

TimMonko replied:
@brisvag. Which should we merge first? I think #7870 since it's more complicated. I'm happy to dig in on this PR over the weekend, if needed

lukasz-migas replied:
I would say merge #7870 first.  If it can be merged before the end of the weekend, I am happy to fix what gets broken 😉 I am on holiday next week but will be back on the 30th.

brisvag replied:
yeah, I agree!

lukasz-migas committed:
Update _viewer_mouse_bindings.py

- improve docstrings

Czaki commented on src/napari/_qt/_tests/test_qt_viewer.py:
```diff
@@ -1381,3 +1381,59 @@ def test_dask_cache(qt_viewer):
     )
+
+
+def test_viewer_drag_to_zoom(qtbot, make_napari_viewer):
```
This test could be wirten using `qt_viewer` and `viewer_model` fixtures. 

Currently, spawning whole viewer is slower than just `qt_viewer` and more fragile

lukasz-migas replied:
Thanks for the feedback, Grzegorz. Indeed, it can - will change it in the next commit.

Czaki commented on src/napari/_qt/layer_controls/qt_layer_controls_base.py:
```diff
@@ -108,6 +109,8 @@ def __init__(self, layer: Layer) -> None:
             extra_tooltip_text=trans._(
                 '\n(or hold Space)\n(hold Shift to pan in 3D)'
+                '\n(hold {control} to zoom via ROI selection)',
+                control=KeyCode.from_string('Alt').os_symbol(),
```
I don't like hardcoded shortcuts. May we add here TODO note that this need to be fixed with app-model migration ad shortcut needs to be read from settings in future?

lukasz-migas replied:
I can change to just the text version `Alt` as it's in the transform button.

Czaki commented on src/napari/_vispy/_tests/test_vispy_zoom_overlay.py:
```diff
@@ -0,0 +1,14 @@
+
+
+def test_zoom_overlay_initialization(make_napari_viewer):
+    viewer = make_napari_viewer()
```
Will it work with passing `ViewerModel` instead of `Viewer`?

lukasz-migas replied:
Yes - and it's around 6 times faster (0.2s vs 1.2s). I think this approach would also work for a couple of other tests in that folder.

Czaki commented on src/napari/components/_tests/test_zoom.py:
```diff
@@ -0,0 +1,22 @@
+    # validate canvas_positions
+    zoom.canvas_positions = ((0, 0), (300, 200))
+    with pytest.raises(ValidationError):
+        # data_positions must be 3-D
```
```suggestion
        # data_positions must be 2-D
```
? 
As it is expected to fail?

lukasz-migas replied:
Yes - just making sure the validator is working as expected.

Czaki commented on src/napari/components/_viewer_mouse_bindings.py:
```diff
@@ -1,5 +1,9 @@
 
+# This is the minimum size of the zoom box (in pixels) that will
+# trigger a zoom when the user drags the mouse while holding Alt.
+MIN_ZOOMBOX_SIZE = 5
```
Looks small for me.

Czaki reviewed (commented):
Minor comments for future maintainability

lukasz-migas committed:
Minor fixes

Czaki commented on src/napari/_qt/_tests/test_qt_viewer.py:
```diff
@@ -1381,3 +1381,58 @@ def test_dask_cache(qt_viewer):
+    canvas = qt_viewer.canvas
+
+    if os.getenv('CI'):
+        viewer.show()
```
```suggestion
        qt_viewer.show()
```

Czaki commented on src/napari/_qt/_tests/test_qt_viewer.py:
```diff
@@ -1381,3 +1381,58 @@ def test_dask_cache(qt_viewer):
     )
+
+
+def test_viewer_drag_to_zoom(qt_viewer, qtbot):
+    """Test drag to zoom mouse binding."""
+    viewer = qt_viewer.viewer
```
The `qt_viewer.viewer` is `ViewerModel` that could be accessed using `viewer_model`   fixture.

lukasz-migas committed:
Update test_qt_viewer.py

TimMonko committed:
Merge branch 'main' into add-zoom-box

lukasz-migas committed:
Update zoom.py

lukasz-migas committed:
Update test_qt_viewer.py

lukasz-migas committed:
Update viewer_model.py

lukasz-migas committed:
Fixes

brisvag committed:
use simpler zoom logic

brisvag committed:
take abs value to account for flipped box

brisvag committed:
handle grid mode correctly

TimMonko committed:
Merge branch 'main' into add-zoom-box

TimMonko committed:
remove utils.zoom tests

TimMonko committed:
revert lint of test_viewer_mouse_bindings
